---
title: "Simulating Whole-Genome Inherited Bisulphite Sequencing Data"
author: Pascal Belleau, Astrid DeschÃªnes and Arnaud Droit 
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Simulating Whole-Genome Inherited Bisulphite Sequencing Data}
  %\VignettePackage{methylInheritanceSim}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo = FALSE, warning=FALSE, message=FALSE, results = 'asis'}
BiocStyle::markdown()
library(knitr)
```

<br />
**Package**: `r Rpackage("methylInheritanceSim")`<br />
**Authors**: `r packageDescription("methylInheritanceSim")[["Author"]]`<br />
**Version**: `r packageDescription("methylInheritanceSim")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("methylInheritanceSim")[["License"]]`<br />


# Licensing 

The `r Rpackage("methylInheritanceSim")` package and the underlying 
`r Rpackage("methylInheritanceSim")` code 
are distributed under the Artistic license 2.0. You are free to use and 
redistribute this software. 


# Introduction

DNA methylation plays an important role in the biology of tissue development 
and diseases. High-throughput sequencing techniques enable genome-wide 
detection of differentially methylated elements (DME), commonly sites (DMS) or 
regions (DMR). The analysis of treatment effects on DNA methylation, from 
one generation to the next (inter-generational) and across generations that 
were not exposed to the initial environment (trans-generational) represent 
complex designs. There are two main approaches to study the methylation 
inheritance, the first 
is based on segregation in pedigree while the second uses the intersection 
between the DME of each generation (useful when pedigree is unknown). The 
power and the false positve rate of those types of design are relatively 
hard to evaluate.

We present a package that simulates the methylation inheritance. Using real 
datasets, the package generates a synthetic chromosome by sampling regions. 
Two different distributions are used to simulate the methylation level at 
each CpG site: one for the DMS and one for all the other sites. The second 
distribution takes advantage of parameters estimated using the control 
datasets. The package also
offers the option to select the proportion of sites randomly fixed as DMS, 
as well as, the fraction of the cases that inherited the DMS in the 
subsequent generations. 

The `r Rpackage("methylInheritanceSim")` package generates simulated 
multigenerational 
DMS datasets that are
useful to evaluate the power and the false discovery rate 
of experiment design analyse, such as the
`r Rpackage("methylInheritance")` package does.


# Loading methylInheritanceSim package

As with any R package, the `r Rpackage("methylInheritanceSim")` package should 
first be loaded with the following command:

```{r loadingPackage, warning=FALSE, message=FALSE}
library(methylInheritanceSim)
```


# Description of the simulation process

The first step of the simulation process is to create a synthetic chromosome 
made up of methylated sites. The synthetic methylated sites (or CpG sites) are 
generated 
using a real dataset (**methData** parameter). The read dataset only needs to
contain methylation for controls on one generation; a real multigenerational 
dataset is not needed. 

Two parameters are critical during this process:

* **nbBlock**: The number of blocks randomly selected in the 
real dataset genome. 
* **lBlock**:  The number of consecutive methylated sites that must contain 
each selected block. 

Those two parameters unable to reproduce CpG islands of customizable size. 
It also reproduces the relation between the methylation level and the
distance associated to adjacent methylated sites.

![alt text](syntheticChr.png "Synthetic Chr")

For each methylated site of the synthetic chromosome, the 
alpha and beta parameters of a Beta distribution are estimated 
from the mean and variance of
the proportion of C/T at the site of the real control dataset. 

## Simulated control dataset

A Beta distribution is
used to simulate the proportion of C/T in the methylated sites of the 
simulated control dataset.

Using the synthetic chromosome, DMS are randomly selected from the methylated 
sites. The **rateDiff** parameter fixes the mean of the proportion sites that 
are differentially methylated (DMS). 

To recreate differentially methylated regions (DMR), the successors site
of a DMS, located within 1000 base pairs, has a 
higher probability to be selected as a DMS.

The inheritance is done through the DMR. This means that when the following 
generation inherits of a DMR region, it inherits all of the DMS present in the
region. The **propInherite** parameter fixes the proportion of DMR 
that are inherited.

## Simulated case dataset 

For the methylated sites in the F1 generation of the 
simulated case dataset, 
a Beta distribution is used to simulate the 
proportion of C/T. This is the same distribution as for the control dataset.

A proportion of the methylated sites, fixed by the **vpDiff** parameter, is 
selected to be DMS. Those DMS are assigned an updated proportion of C/T that 
follows a shifted
Beta distribution with parameters estimated using the mean 
of control $\pm$ **vDiff**. The **vpDiff** parameter is similar 
to penetrance. 

In the subsequent generation, a proportion of the DMS present in the simulated
case dataset are selected to be
inherited. The proposition of inherited DMS is calculated as:

$$ \mathbf{vpDiff\   \times\   {vInheritance}^{number\   of\  generations\  after\  F2} }$$

The proportion of C/T of those selected inherited sites follows a shifted 
Beta distribution with parameters estimated using mean of 
control $\pm$ (**vDiff** x**propHetero**).
The **propHetero** is 0.5 if one of the parent is a control.


# Case study 

## The dataset

A dataset containing methylation data (6 cases and 6 controls) has been 
generated using the `r Rpackage("methylInheritanceSim")` package using a real 
dataset from Rat experiment (the real dataset is not public yet, so we used a 
simulation based on it). The data have been formated, using  
the `r Rpackage("methylkit")` package, into a *methylBase* object 
(using the `r Rpackage("methylkit")` functions: *filterByCoverage*, 
*normalizeCoverage* and *unite*).

```{r caseStudy01, warning=FALSE, message=FALSE, collapse=TRUE}
## Load  read DMS dataset (not in this case but normaly)
data(samplesForChrSynthetic)

## Print the first three rows of the object
head(samplesForChrSynthetic, n = 3)
```


## The simulation

The simulation is run using the **runSim** function.
The **outputDir** parameter fixes the directory where the results are stored.


```{r runSim01, warning=FALSE, message=FALSE, collapse=TRUE, cache=TRUE}

temp_dir <- "test_runSim"

runSim(outputDir = temp_dir, 
        fileID = "S1", 
        nbSynCHR = 1, # The number of synthetic chromosome
        methData = samplesForChrSynthetic,  # The dataset use for generate 
                                            # the synthetic chr.
        nbSimulation = 2,  # The number of simulation for each parameters
        nbBlock = 10, nbCpG = 20, # The number of site in the 
                                # synthetic chr is nbBLock * nbCpG
        nbGeneration = 3, 
        vNbSample = c(3, 6), 
        vpDiff = c(0.9), 
        vpDiffsd = c(0.1), 
        vDiff = c(0.8), # The shift of the mean of the differentially 
                        # methylated sites
        vInheritance = c(0.5), propInherite = 0.3,
        rateDiff = 0.3, # The mean frequency of the differentially 
                        # methylated region TODO : region ?
        minRate = 0.2, 
        propHetero = 0.5, 
        runAnalysis = TRUE, 
        nbCores = 1, 
        vSeed = 32)         # Fix seed to unable reproductive results
        
        # The files generated
        dir(temp_dir)
```
        
```{r removeFiles, warning=FALSE, message=FALSE, collapse=TRUE, echo=FALSE}       
if (dir.exists(temp_dir)) {
    unlink(temp_dir, recursive = TRUE, force = FALSE)
}
```


# Files generated by the simulation

Three files type are generated by default: 

The first file contains information about the synthetic chromosome.
This is a GRange that contains the CpG (or methylated sites).
The file has four metadata from the real dataset:
chrOri the chromosome from the real dataset
startOri the position of the site in the real dataset
meanCTRL the mean of the control in the real dataset
varCTRL the variance of the control in the real dataset.

The file name is 
composed of those elements, separated by "_":
stateInfo, 
a code for the simulation (the fileId parameter, ex S1), 
the chromosome number, 
the file extension ".rds".
Ex : stateInfo_S1_1.rds


The second file is a GRange contains information about the simulation.
The file has four metadata related to real dataset:
meanDiff the mean of the shifted distribution
meanCTRL.meanCTRL the mean of the control distribution
partitionCase number of cases simulated with the shifted distribution
partitionCtrl number of cases simulated with the control distribution
and a metadata for each case and controls the proportion of C/T.


The third file is a list with 2 entries. The first entry is called 
stateDiff and contains a vector of integers (0 and 1) with 
a length corresponding the length of stateInfo. The statDiff
indicates, using a 1, the positions where the CpG sites are
differentially methylated. The second entry is
called statInherite and contains a vector of integers 
(0 and 1)
with a length corresponding the length of stateInfo. The 
statInherite
indicates, using a 1, the positions where the CpG values are
inherited.

The package generate a list of files depending on the parameters :
saveGRanges, saveMethylKit and, runAnalysis. 

When saveGRanges parameters is TRUE, the package save two files type. 
The first generate for each simulation contains a list. The length of the list 
corresponds to the number of generations. The generations are stored in order 
(first entry = first generation, second entry = second generation, etc..). 
All samples related to one generations are contained in a GRangesList. 
The GRangeaList store a list of GRangea. Each GRanges stores the raw 
methylation data of one sample.
The second file a numeric vector denoting controls and cases (a file is 
generated by entry in the vector parameters vNbSample).

When **saveMethylKit** is **TRUE**, for each simulation save a file contains a list.
The length of the list corresponds to the number of generations. The generations 
are stored in order (first entry = first generation, second entry = second 
generation, etc..). All samples related to one generation are contained in a 
S4 methylRawList object. The methylRawList object contains two Slots:

1. **treatment**: A numeric vector denoting controls and cases.
2. **.Data**: A list of methylRaw objects. Each object stores the raw 
methylation data of one sample.

When **runAnalysis** is **TRUE**, two files are saved for each simulation:

* The first file is the methylObj... file formated with 
the `r Rpackage("methylkit")` package in a **S4 methylBase** object 
(with the methylKit functions: filterByCoverage, normalizeCoverage and unite).
* The second file contains a **S4 calculateDiffMeth** object generates with 
the `r Rpackage("methylKit")` functions calculateDiffMeth on the first file.


# Conclusion

The `r Rpackage("methylInheritanceSim")` package generates simulated 
multigenerational DMS datasets. Several simulator parameters can be derived 
from real dataset provided by the user in order to replicate realistic 
case-control scenarios. 

The results of a simulation could be analysed, using the 
`r Rpackage("methylInheritance")` package, to evaluate the power and the 
false discovery rate of an experiment design.






