---
title: "Whole-Genome Inherited Bisulphite Sequencing Data Simulation"
author: Pascal Belleau, Astrid Deschenes and Arnaud Droit 
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{TODO}
  %\VignettePackage{methylInheritanceSim}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
library(knitr)
```

<br />
**Package**: `r Rpackage("methylInheritanceSim")`<br />
**Authors**: `r packageDescription("methylInheritanceSim")[["Author"]]`<br />
**Version**: `r packageDescription("methylInheritanceSim")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("methylInheritanceSim")[["License"]]`<br />


# Licensing 

The **methylInheritanceSim** package and the underlying 
**methylInheritanceSim** code 
are distributed under the Artistic license 2.0. You are free to use and 
redistribute this software. 

# Introduction

# Introduction

DNA methylation plays an important role in the biology of tissue development 
and diseases. High-throughput sequencing techniques enable genome-wide 
detection of differentially methylated elements (DME), commonly sites (DMS) or 
regions (DMR). The analysis of treatment effects on DNA methylation, from 
one generation to the next (inter-generational) and across generations that 
were not exposed to the initial environment (trans-generational) represent 
complex designs. There are two main approaches to study the methylation inheritance, the first 
is based on segregation in pedigree while the second uses the intersection 
between the DME of each generation (useful when pedigree is unknown). The power 
and the false positve rate of this type of design are relatively hard to 
evaluate.

We present a package simulating the methylation inheritance. Using real 
datasets, the method generates a synthetic chromosome by sampling regions. Two 
different distributions are used to simulate the methylation level at each CpG 
site: one for the DMS and one for all the other sites. The second distribution 
takes advantage of parameters estimated using control datasets. The package also
offers the option to select the proportion of sites randomly selected as DMS, 
as well as, the fraction of the cases that inherited the DMS in the subsequent 
generations.

# Loading methylInheritanceSim package

As with any R package, the `r Rpackage("methylInheritanceSim")` package should 
first be loaded with the following command:

```{r loadingPackage, warning=FALSE, message=FALSE}
library(methylInheritanceSim)
```


# Description of the permutation analysis

The synthetic methylated sites (or CpG sites) are generate using a real dataset (methData). Two parameters
determine this step (nbBlock and lBLock). The nbBlock representes the number 
block randomly selected in the real dataset genome. Each block must contain 
lBlock consecutive methylated sites.

![alt text](syntheticChr.png "Synthetic Chr")

For each methylated site of the synthetic chromosome, we estimate the 
parameters alpha and beta of a distribution beta from the mean and variance of
the proportion of C/T  for the CTRL at this site. We use this Beta distribution 
to simulate the the proportion of C/T in the controle dataset.

We select randomly the DMS from methylated sites from synthetic chromosome 
(mean of the proportion of DMS on the number of methylated sites is rateDiff). To 
create DMR, the successors of the selected DMS are more likely to be DMS too, 
if they are nearer than 1000 base pairs. After a proportion (propInherite) of 
the DMR are selected has inherited 

For the case simulation of the F1 generation, for methilated sites not selected 
as DMS, the same beta distribution as the control is used to simulate the 
proportion of C/T. For the DMS, a proportion (vpDiff) of the cases are selected 
and the C/T follow a shift Beta 
distribution (the parameter are estimated with mean of control + or - vDiff). 
The 1 - vpDiff other case follow the same distribution as a control. 
The parameter vpDiff is liked penetrance. 

For the subsequent generation, for the selected DMS inherited, a proportion of 
the case (vpDiff x power(vInheritance, number of generation after F2) ). The 
proportion of C/T of the selected cases follow a shifted Beta distribution 
(the parameter are estimated with mean of control + or - (vDiff * propHetero) ).
The propHetero is 0.5 if one of the parent are control


# OutputDir

Three files 

The first file contain information about the synthetic chromosome.
The file name is 
composed of those elements, separated by "_":
stateInfo
a code for the simulation (ex S1)
the chromosome number.

The is GRange that contains the CpG (or methylated sites).
The file have four metadata from the real dataset:
chrOri the chromosome from the real dataset
startOri the position from the real dataset
meanCTRL the mean of the control in the real dataset
varCTRL the variance of the control in the real dataset.

The second file is a GRange contains information about the simulation
The file have four metadata from the real dataset
meanDiff means of the shifted distribution
meanCTRL.meanCTRL means of the control distribution
partitionCase number of case simulated with the shifted distribution
partitionCtrl number of case simulated with the control distribution
number of controls ctrl.V? The proportion of C/T for this case
number of cases case.V? THe proportion of C/T for this case


The third file stateDiff is a list with 2 entries. The first entry is called 
stateDiff and contains a vector of integer (0 and 1) with 
a length corresponding the length of stateInfo. The statDiff
indicates, using a 1, the positions where the CpG sites are
differentially methylated. The second entry is
called statInherite and contains a vector of integer 
(0 and 1)
with a length corresponding the length of stateInfo. The 
statInherite
indicates, using a 1, the positions where the CpG values are
inherited.

The package generate a list of file depending of the parameters :
saveGRanges, saveMethylKit and, anaMethylKit. 

If saveGRanges parameters is TRUE, for each simulations 
The file name is 
composed of those elements, separated by "_":
methylGR and treatment
fileID
the chromosome number

the number of samples

the mean proportion of case samples that will have,
for a specific position, differentially methylated values

the proportion of C/T for a case differentially methylated 
follow a shifted beta distribution, a number in the vDiff vector

the proportion of cases that inherited differentially sites, a number 
in the vInheritance vector

Id for the simulation, a number 
between 1 and nbSimulation

the file extension ".rds"

